src/strat_engine/blockdev.rs

```diff
@@ -0,0 +1,373 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::io::{Read, Write, ErrorKind, Seek, SeekFrom};
+use std::fs::{OpenOptions, read_dir};
+use std::path::{Path, PathBuf};
+use std::io;
+use std::str::{FromStr, from_utf8};
+use std::cmp::Ordering;
+
+use nix::sys::stat;
+use time::Timespec;
+use devicemapper::Device;
+use crc::crc32;
+use byteorder::{LittleEndian, ByteOrder};
+use uuid::Uuid;
+use bytesize::ByteSize;
+
+use types::{Sectors, SectorOffset};
+use engine::{EngineResult, EngineError};
+
+use consts::*;
+use super::consts::*;
+use super::util::blkdev_size;
+
+pub use super::BlockDevSave;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct MDA {
+    pub last_updated: Timespec,
+    used: u32,
+    length: u32,
+    crc: u32,
+    offset: SectorOffset, // From start of MDA, not BDA
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct BlockDev {
+    pub stratis_id: String,
+    pub dev: Device,
+    pub id: String,
+    pub path: PathBuf,
+    pub sectors: Sectors,
+    pub mdaa: MDA,
+    pub mdab: MDA,
+    mda_sectors: Sectors,
+    reserved_sectors: Sectors,
+}
+
+impl BlockDev {
+    pub fn new(stratis_id: &str,
+               path: &Path,
+               mda_size: Sectors,
+               force: bool)
+               -> EngineResult<BlockDev> {
+        let pstat = try!(stat::stat(path));
+
+        if pstat.st_mode & 0x6000 != 0x6000 {
```

Comment:
support for this is now in stdlib (https://doc.rust-lang.org/std/os/unix/fs/trait.FileTypeExt.html) so we can use that.